src/Lecture4.hs

```diff
@@ -134,8 +148,43 @@ errors. We will simply return an optional result here.
 🕯 HINT: Use the 'readMaybe' function from the 'Text.Read' module.
 -}
 
+split :: Char -> String -> [String]
+split symb = go (== symb)
+  where go :: (Char -> Bool) -> String -> [String]
+        go isSymb str = case dropWhile isSymb str of
+                          "" -> []
+                          subStr -> term : go isSymb rest
+                                      where (term, rest) = break isSymb subStr
+
+validateRow :: [String] -> Maybe (String, String, Int)
+validateRow [name, tradeType, cost] = if isNameValid name
+                                         && isTradeTypeValid parsedTradeType
+                                         && isCostValid parsedCost
+                                        then Just (name, parsedTradeType, read parsedCost :: Int)
+                                        else Nothing
+  where parsedTradeType :: String
+        parsedTradeType = dropSpaces tradeType
+
+        parsedCost :: String
+        parsedCost = dropSpaces cost
+    
+        isNameValid :: String -> Bool
+        isNameValid = (/= "")
+
+        isTradeTypeValid :: String -> Bool
+        isTradeTypeValid trType = trType == show Sell || trType == show Buy
+
+        isCostValid :: String -> Bool
+        isCostValid cst = maybe False (>= 0) (readMaybe cst :: Maybe Int)
+validateRow _                       = Nothing
+
+toRow :: (String, String, Int) -> Row
+toRow (name, tradeType, cost) = Row { rowProduct   = name
+                                    , rowTradeType = if tradeType == show Sell then Sell else Buy
+                                    , rowCost      = cost }
+
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow = fmap toRow . validateRow . split ','
```

Comment:
You have an excellent implementation using composition of functions 🔝 

I haven't explained the following during lectures but there's a nice way to write this parsing syntax using `do`-notation, `Monad` instance for `Maybe` and the `MonadFail` desugaring trick 🧙🏻 

You can read more about this trick here:

* https://kowainik.github.io/posts/haskell-mini-patterns#monadfail-sugar

Applying this programming pattern and a few helper functions, the `parseRow` function will look like this:

```suggestion
parseRow str = do
    [rawProd, rawTrade, rawCost] <- Just (split ',' str)
    guard $ rawProd /= ""
    trade <- readMaybe rawTrade
    cost <- readMaybe rawCost
    guard $ cost >= 0
    pure $ Row rawProd trade cost
```